doc/reshaping.rst

```diff
@@ -133,6 +133,38 @@ pandas, it does not automatically drop missing values. Compare:
 We departed from pandas's behavior here because predictable shapes for new
 array dimensions is necessary for :ref:`dask`.
 
+Stacking different variables together
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+These stacking and unstacking operations are particularly useful for reshaping
+xarray objects for use in machine learning packages, such as `scikit-learn
+<http://scikit-learn.org/stable/>`_, that usually require two-dimensional numpy
+arrays as inputs. For datasets with only one variable, we only need ``stack``
+and ``unstack``, but combining multiple variables in a
+:py:class:`xarray.Dataset` is more complicated. If the variables in the dataset
+have matching numbers of dimensions, we can call
+:py:meth:`~xarray.Dataset.to_array` and then stack along the the new coordinate.
+But :py:meth:`~xarray.Dataset.to_array` will broadcast the dataarrays together,
+which will effectively tile the lower dimensional variable along the missing
+dimensions. The method :py:meth:`xarray.Dataset.to_stacked_array` allows
+combining variables of differing dimensions without this wasteful copying while
+:py:meth:`xarray.DataArray.to_unstacked_dataset` reverses this operation.
+Just as with :py:meth:`xarray.Dataset.stack` the stacked coordinate is
+represented by a :py:class:`pandas.MultiIndex` object. These methods are used
+like this:
+
+.. ipython:: python
+
+        arr = xr.DataArray(np.arange(6).reshape(2, 3),
+                        coords=[('x', ['a', 'b']), ('y', [0, 1, 2])])
+        data = xr.Dataset({'a': arr, 'b': arr.isel(y=0)})
```

Comment:
I was a bit confused by the labels 'a' et 'b' used for both the names of the data variables and the 'x' coordinate labels in this example (and also in the docstrings). As the stack operation is rather complex here, maybe a slightly simpler example would help users better understand it? Something like:

```python
>>> data = xr.Dataset(
...     data_vars={'a': (('x', 'y'), [[0, 1, 2], [3, 4, 5]]),
...                'b': ('x', [6, 7])},
...     coords={'y': ['u', 'v', 'w']}
... )
<xarray.Dataset>
Dimensions:  (x: 2, y: 3)
Coordinates:
  * y        (y) <U1 'u' 'v' 'w'
Dimensions without coordinates: x
Data variables:
    a        (x, y) int64 0 1 2 3 4 5
    b        (x) int64 6 7

>>> data.to_stacked_array('z', ['y'])
<xarray.DataArray (x: 2, z: 4)>
array([[0, 1, 2, 6],
       [3, 4, 5, 7]])
Coordinates:
  * z         (z) MultiIndex
  - variable  (z) object 'a' 'a' 'a' 'b'
  - y         (z) object 'u' 'v' 'w' nan
Dimensions without coordinates: x
```

Reply:
+1
A nice sketch in the doc would be illuminating

Reply:
I am fine making this change. Not sure how to draw the sketch though.

Reply:
Okay. I used your example in `reshaping.rst` and in `to_stacked_array`.